akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -204,14 +226,107 @@ object ActorGraphInterpreter {
       override def toString: String = BatchingActorInputBoundary.this.toString
     })
 
-    override def toString: String = s"BatchingActorInputBoundary(id=$id, fill=$inputBufferElements/$size, completed=$upstreamCompleted, canceled=$downstreamCanceled)"
+    override def toString: String = s"BatchingActorInputBoundary(forPort=$internalPortName, fill=$inputBufferElements/$size, completed=$upstreamCompleted, canceled=$downstreamCanceled)"
+  }
+
+  final case class SubscribePending(boundary: ActorOutputBoundary) extends SimpleBoundaryEvent {
+    override def execute(): Unit = boundary.subscribePending()
+
+    override def shell: GraphInterpreterShell = boundary.shell
+
+    override def logic: GraphStageLogic = boundary
+  }
+
+  final case class RequestMore(boundary: ActorOutputBoundary, demand: Long) extends SimpleBoundaryEvent {
+    override def execute(): Unit = {
+      if (GraphInterpreter.Debug) println(s"${boundary.shell.interpreter.Name}  request  $demand port=${boundary.internalPortName}")
+      boundary.requestMore(demand)
+    }
+    override def shell: GraphInterpreterShell = boundary.shell
+    override def logic: GraphStageLogic = boundary
+  }
+  final case class Cancel(boundary: ActorOutputBoundary) extends SimpleBoundaryEvent {
+    override def execute(): Unit = {
+      if (GraphInterpreter.Debug) println(s"${boundary.shell.interpreter.Name}  cancel port=${boundary.internalPortName}")
+      boundary.cancel()
+    }
+
+    override def shell: GraphInterpreterShell = boundary.shell
+    override def logic: GraphStageLogic = boundary
+  }
+
+  private[stream] class OutputBoundaryPublisher(boundary: ActorOutputBoundary, internalPortName: String) extends Publisher[Any] {
```

Comment:
Check if the old `BoundaryPublisher` is still needed. There might be some actor backed publisher implementations around that can be removed, please investigate.